File: Protocol/_theoreme_sim/src/propagation/neighborChargePropagation.js
/* Purpose: Simulate asynchronous neighbor charge learning with max-based estimate updates. */

import {
  getLinkDeliveryProbability,
  recordLinkUsage,
} from "./linkUsageTracker.js";

/**
 * @param {{
 *  edges:Array<{a:number,b:number}>,
 *  nodes:Map<number,{qTotal:number}>,
 *  estimates:Map<number,Map<number,number>>,
 *  config:{deliveryProbability:number},
 *  rng:() => number
 * }} state
 * @returns {{updates:number, deliveries:number}}
 */
export function propagateNeighborChargesRound(state) {
  let updates = 0;
  let deliveries = 0;

  const tryDeliver = (observerId, senderId) => {
    const probability = getLinkDeliveryProbability(state, observerId, senderId);
    if (state.rng() > probability) {
      return;
    }

    recordLinkUsage(state, observerId, senderId, 0.35);
    deliveries += 1;
    const estimateMap = state.estimates.get(observerId);
    const oldValue = estimateMap.get(senderId) ?? 0;
    const advertisedCharge = state.nodes.get(senderId).qTotal;
    const nextValue = Math.max(oldValue, advertisedCharge);

    if (nextValue > oldValue) {
      updates += 1;
      estimateMap.set(senderId, nextValue);
    }
  };

  for (const edge of state.edges) {
    tryDeliver(edge.a, edge.b);
    tryDeliver(edge.b, edge.a);
  }

  return { updates, deliveries };
}

File: Protocol/_theoreme_sim_py/src/propagation/neighbor_charge_propagation.py
"""Purpose: Asynchronous-like neighbor charge learning with max-based estimate updates."""

from __future__ import annotations

from src.propagation.link_usage_tracker import (
    get_link_delivery_probability,
    record_link_usage,
)


def propagate_neighbor_charges_round(state: dict) -> dict:
    updates = 0
    deliveries = 0

    def try_deliver(observer_id: int, sender_id: int) -> None:
        nonlocal updates, deliveries

        probability = get_link_delivery_probability(state, observer_id, sender_id)
        if state["rng"]() > probability:
            return

        record_link_usage(state, observer_id, sender_id, 0.35)
        deliveries += 1

        estimate_map: dict[int, float] = state["estimates"].get(observer_id, {})
        old_value = float(estimate_map.get(sender_id, 0.0))
        advertised_charge = float(state["nodes"][sender_id].q_total)
        next_value = max(old_value, advertised_charge)

        if next_value > old_value:
            updates += 1
            estimate_map[sender_id] = next_value

    for edge in state["edges"]:
        try_deliver(edge["a"], edge["b"])
        try_deliver(edge["b"], edge["a"])

    return {"updates": updates, "deliveries": deliveries}
